### PR TITLE
Merkle proof documentation

### PR DESCRIPTION
**Pull request by simonasstipinas**

### CODE REVIEW

File: utils/merkle_proof/src/lib.rs [hunk 1]
```diff
@@ -1,129 +1,137 @@
...
-    pub fn left_and_right_branches(&self) -> Option<(&Self, &Self)> {
-        match *self {
-            MerkleTree::Leaf(_) | MerkleTree::Zero(0) => None,
-            MerkleTree::Node(_, ref l, ref r) => Some((l, r)),
-            MerkleTree::Zero(depth) => Some((&ZERO_NODES[depth - 1], &ZERO_NODES[depth - 1])),
-        }
-    }
+fn get_generalized_index_bit(index: usize, position: usize) -> bool {
+    // ((index >> position) & 0x01) == 1 lighthouse 
+    ((index >> position) & 0x01) > 0 //dokumentacija
```
Comment (faustuzas) [inline utils/merkle_proof/src/lib.rs]: please remove Lithuanian comments

File: utils/merkle_proof/Cargo.toml [hunk 2]
```diff
@@ -5,12 +5,16 @@ authors = ["simonass <simonas.stipinas@programmatus.com>"]
...
-
 [dependencies]
 ethereum-types = "0.6"
 eth2_hashing = { path = "../eth2_hashing" }
 lazy_static = "1.3.0"
+eth2_ssz_types = { path = "../ssz_types" }
+eth2_ssz = { path = "../ssz" }
+
+[lib]
+name = "math"
```
Comment (faustuzas) [inline utils/merkle_proof/Cargo.toml]: your library name is for sure not "math"

File: utils/merkle_proof/Cargo.toml [hunk 3]
```diff
@@ -5,12 +5,16 @@ authors = ["simonass <simonas.stipinas@programmatus.com>"]
...
 
-[dev-dependencies]
-quickcheck = "0.8"
-quickcheck_macros = "0.8"
\ No newline at end of file
+# [dev-dependencies]
+# quickcheck = "0.8"
+# quickcheck_macros = "0.8"
```
Comment (faustuzas) [inline utils/merkle_proof/Cargo.toml]: if dependency is redundant, remove it

File: utils/merkle_proof/src/lib.rs [hunk 4]
```diff
@@ -1,129 +1,137 @@
...
+}
 
-    pub fn is_leaf(&self) -> bool {
-        match self {
-            MerkleTree::Leaf(_) => true,
-            _ => false,
-        }
-    }
+fn generalized_index_sibling(index: usize) -> usize {
+    return index^1;
```
Comment (faustuzas) [inline utils/merkle_proof/src/lib.rs:65]: This is really the power of 1? Why not just return index then?

File: utils/merkle_proof/src/lib.rs [hunk 5]
```diff
@@ -1,129 +1,137 @@
...
 
-        proof.reverse();
+fn generalized_index_parent(index: usize) -> usize {
+    return index / 2;
+}
 
-        (current_node.hash(), proof)
+fn get_branch_indices(tree_index: usize) -> Vec<usize> {
+
+    let mut o = vec![generalized_index_sibling(tree_index)];
```
Comment (faustuzas) [inline utils/merkle_proof/src/lib.rs]: please use more meaningful names

File: utils/merkle_proof/src/lib.rs [hunk 6]
```diff
@@ -133,66 +141,132 @@ fn calculate_merkle_root(
...
+    return *btree_first.get(&1usize).unwrap();    
 }
 
-//paklausti ar reik:
-// kokiu funkciju man reik?-nera dokumentacijoj;
-// verify_merkle_multiproof? yra dokumentacijoj
+
+
+//TEST
+//-----------------------------------------
```
Comment (faustuzas) [inline utils/merkle_proof/src/lib.rs]: This comment brings no value, we can clearly see that items below are tests

File: utils/merkle_proof/src/lib.rs [hunk 7]
```diff
@@ -133,66 +141,132 @@ fn calculate_merkle_root(
...
+        assert_eq!(5, concat_generalized_indices(&general_indices));
+    }
 
-        let root = hash_and_concat(node_b0x, node_b1x);
+    #[test]
+    fn get_generalized_index_length_test() {
+        assert_eq!(get_generalized_index_length(4),2);
+        assert_eq!(get_generalized_index_length(7),2);
+        assert_eq!(get_generalized_index_length(9),3);
+        // assert_eq!(get_generalized_index_length(0b00),0);
```
Comment (faustuzas) [inline utils/merkle_proof/src/lib.rs]: remove uncommented code

File: utils/merkle_proof/src/lib.rs (hunk 4 above)
Comment (simonasstipinas) [reply utils/merkle_proof/src/lib.rs:65]: This is XOR operator.

File: utils/merkle_proof/src/lib.rs [hunk 8]
```diff
@@ -1,129 +1,149 @@
...
+    HashSet::from_iter(data.iter().cloned())
+}
 
-    assert_eq!(proof.len(), depth, "proof length should equal depth");
+fn verify_merkle_proof(leaf: H256, proof: &[H256], index: usize, root: H256) -> Result<bool, MerkleProofError> {
+    match calculate_merkle_root(leaf, proof, index) {
+        Ok(calculated_root) => Ok(calculated_root== root),
+        Err(err) => Err(err),
+    }
+    // return calculate_merkle_root(leaf, proof, index) == root
```
Comment (faustuzas) [inline utils/merkle_proof/src/lib.rs]: Remove commented code

File: utils/merkle_proof/src/lib.rs [hunk 9]
```diff
@@ -1,129 +1,149 @@
...
+}
 
+fn calculate_merkle_root(leaf: H256, proof: &[H256], index: usize) -> Result<H256, MerkleProofError> {
+    // assert_eq!( proof.len(), get_generalized_index_length(index), "Length of proof should equal generalized index depth");
```
Comment (faustuzas) [inline utils/merkle_proof/src/lib.rs]: remove

File: utils/merkle_proof/src/lib.rs [hunk 10]
```diff
@@ -133,67 +153,167 @@ fn calculate_merkle_root(
             root = hash(&input);
         }
     }
-
-    H256::from_slice(&root)
+    Ok(H256::from_slice(&root))
+    // return H256::from_slice(&root);
```
Comment (faustuzas) [inline utils/merkle_proof/src/lib.rs]: remove

File: utils/merkle_proof/src/lib.rs [hunk 11]
```diff
@@ -133,67 +153,167 @@ fn calculate_merkle_root(
...
+    let mut keys: Vec<usize> = vec![];
+
+    for(key, _value) in btree_first.iter_mut() {
+        keys.push(key.clone());
+    }
+
+    keys.sort();
+    keys = reverse_vector(keys);
+
+    let mut biggest: usize = *keys.get(0usize).clone().unwrap();
```
Comment (faustuzas) [inline utils/merkle_proof/src/lib.rs:213]: Are you sure you can safely unwrap here?

File: utils/merkle_proof/src/lib.rs [hunk 12]
```diff
@@ -133,67 +153,167 @@ fn calculate_merkle_root(
...
+            keys.push(biggest);
+        }
+        biggest -=1;
+    }
+    keys.sort();
+    keys = reverse_vector(keys);
+    let mut position = 1usize;
+
+    while position < keys.len() {
+        let k = keys.get(position).unwrap();
```
Comment (faustuzas) [inline utils/merkle_proof/src/lib.rs:229]: Same question as up above. If you are using unwrap not in the tests there is a convention to leave a comment why you think it's safe

File: utils/merkle_proof/src/lib.rs [hunk 13]
```diff
@@ -204,77 +324,153 @@ mod tests {
...
+        let leaf_b10 = H256::from([0xCC; 32]);
+        let leaf_b11 = H256::from([0xDD; 32]);
+
+        let node_b0x = hash_and_concat(leaf_b00, leaf_b01);
+        let node_b1x = hash_and_concat(leaf_b10, leaf_b11);
+
+        let root = hash_and_concat(node_b0x, node_b1x);
+        
+
+
```
Comment (faustuzas) [inline utils/merkle_proof/src/lib.rs:407]: why so many whitespaces? :D

**Comment by faustuzas** [comment]
Why there is no public functions in you lib.rs file? At the moment none could use any of your work

**Review by faustuzas — APPROVED** [review]

**Pull request merged**